src/review/logic.py

```diff
@@ -88,9 +205,146 @@ def handle_review_file(_file, content_type, review_assignment, kind):
+            form=submission.review_form,
+        ).order_by('order')
+
+    if review_assignment.results:
+        data = json.loads(review_assignment.results.data)
+
+        for relation in relations:
+            field_name = core_logic.ascii_encode(relation.element.name)
+            v = data[field_name]
+            document.add_heading(relation.element.name, level=1)
+            text = BeautifulSoup(
+                (v[0]).encode('utf-8'),
+                'html.parser'
+            ).get_text()
+            document.add_paragraph(text).bold = True
+            recommendations = {
+                'accept': 'Accept',
+                'reject': 'Reject',
+                'revisions': 'Revisions Required',
+            }
+
+        document.add_heading("Recommendation", level=1)
+        if recommendations.get(review_assignment.recommendation):
+            document.add_paragraph(
+                recommendations[review_assignment.recommendation]
+            ).italic = True
+        document.add_heading("Competing Interests", level=1)
+        document.add_paragraph(
+            review_assignment.competing_interests
+        ).italic = True
+
+    else:
+        p = document.add_paragraph(
+            'You should complete this form and then use the review assignment '
+            'page to upload it.'
+        )
+
+        for relation in relations:
+
+            if (
+                relation.element.field_type
+                in ['text', 'textarea', 'date', 'email']
+            ):
+                document.add_heading(
+                    relation.element.name + ": _______________________________",
+                    level=1
+                )
+                document.add_paragraph(relation.help_text).italic = True
+
+            if relation.element.field_type in ['select', 'check']:
+                document.add_heading(relation.element.name, level=1)
+
+                if relation.element.field_type == 'select':
+                    choices = render_choices(relation.element.choices)
+                else:
+                    choices = ['Y', 'N']
+
+                p = document.add_paragraph(relation.help_text)
+                p.add_run(
+                    ' Mark your choice however you like, '
+                    'as long as it is clear.'
+                ).italic = True
+                table = document.add_table(rows=2, cols=len(choices))
+                hdr_cells = table.rows[0].cells
+
+                for i, choice in enumerate(choices):
+                    hdr_cells[i].text = choice[0]
+
+                table.style = 'TableGrid'
+
+    document.add_page_break()
+
+    if not os.path.exists(os.path.join(settings.BASE_DIR, 'files', 'forms')):
+        os.makedirs(os.path.join(settings.BASE_DIR, 'files', 'forms'))
+
+    path = os.path.join(settings.FORM_DIR, f'{uuid4()}.docx')
+
+    with default_storage.open(path, 'wb') as file_stream:
+        document.save(file_stream)
```

Comment:
Is this correct? `file_stream` is being opened with `'wb'` yet its contents are only being saved to another file?

Reply:
I double-checked and this is correct. The document object will write the docx bytes to the file stream. Python-docx's Document.save method takes a file-like-object as well as a file path as an argument.

https://python-docx.readthedocs.io/en/latest/user/documents.html#opening-a-file-like-document